share/termcap/termcap.5
.\" Copyright (c) 1985, 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)termcap.5	6.11 (Berkeley) 3/6/93
.\"	$Id: termcap.5,v 1.1 1995/10/18 08:44:48 deraadt Exp $
.\"
.Dd March 6, 1993
.Dt TERMCAP 5
.Os BSD 3
.Sh NAME
.Nm termcap
.Nd terminal capability data base
.Sh SYNOPSIS
.Nm termcap
.Sh DESCRIPTION
The
.Nm Termcap
file
is a data base describing terminals,
used,
for example,
by
.Xr \&vi 1
and
.Xr curses 3 .
Terminals are described in
.Nm termcap
by giving a set of capabilities that they have and by describing
how operations are performed.
Padding requirements and initialization sequences
are included in
.Nm termcap .
.Pp
Entries in
.Nm termcap
consist of a number of `:'-separated fields.
The first entry for each terminal gives the names that are known for the
terminal, separated by `|' characters.
The first name given is the most common abbreviation for the terminal.
The last name given should be a long name fully identifying the terminal,
and all others are understood as synonyms for the terminal name.
All names but the last should be in lower case and contain no blanks;
the last name may well contain upper case characters and blanks for
readability.
.Pp
Terminal names (except for the last, verbose entry)
should be chosen using the following conventions.
The particular piece of hardware making up the terminal
should have a root name chosen, thus
.Dq hp2621
This name should not contain hyphens.
Modes that the hardware can be in
or user preferences
should be indicated by appending a hyphen and an indicator of the mode.
Therefore, a
.Dq vt100
in 132-column mode would be
.Dq vt100-w .
The following suffixes should be used where possible:
.Pp
.Bd -filled -offset indent
.Bl -column indent "With automatic margins (usually default)xx"
.Sy Suffix	Meaning	Example
-w	Wide mode (more than 80 columns)	vt100-w
-am	With automatic margins (usually default)	vt100-am
-nam	Without automatic margins	vt100-nam
.Pf \- Ar n Ta No "Number of lines on screen	aaa-60"
-na	No arrow keys (leave them in local)	concept100-na
.Pf \- Ar \&np Ta No "Number of pages of memory	concept100-4p"
-rv	Reverse video	concept100-rv
.El
.Ed
.Sh CAPABILITIES
The characters in the
The
.Em Notes
function
field in the table have the following meanings
(more than one may apply to a capability):
.Pp
.Bd -unfilled -offset indent
N	indicates numeric parameter(s)
P	indicates that padding may be specified
*	indicates that padding may be based on the number of lines affected
o	indicates capability is obsolete
.Ed
.Pp
.Dq Obsolete
capabilities have no
.Em terminfo
equivalents,
since they were considered useless,
or are subsumed by other capabilities.
New software should not rely on them at all.
.Pp
.Bl -column indent indent indent
.Sy Name	Type	Notes	Description
functions
.It "ae	str	(P)	End alternate character set."
.It AL	str	(NP*)	Add"
.Em n
new blank lines
.It "al	str	(P*)	Add new blank line."
.It "am	bool		Terminal has automatic margins."
.It "as	str	(P)	Start alternate character set."
.It "bc	str	(o)	Backspace if not."
.Sy \&^H .
.It "bl	str	(P)	Audible signal (bell)."
.It "bs	bool	(o)	Terminal can backspace with"
.Sy \&^H .
.It "bt	str	(P)	Back tab."
.It "bw	bool	" Ta Sy \&le
(backspace) wraps from column 0 to last column.
.It "CC	str		Terminal settable command character in prototype."
.It "cd	str	(P*)	Clear to end of display."
.It "ce	str	(P)	Clear to end of line."
.It "ch	str	(NP)	Set cursor column (horizontal position)."
.It "cl	str	(P*)	Clear screen and home cursor."
.It "CM	str	(NP)	Memory-relative cursor addressing."
.It "cm	str	(NP)	Screen-relative cursor motion."
.It "co	num		Number of columns in a line (See"
.Sx BUGS
section below).
.It "cr	str	(P)	Carriage return."
.It "cs	str	(NP)	Change scrolling region (VT100)."
.It "ct	str	(P)	Clear all tab stops."
.It "cv	str	(NP)	Set cursor row (vertical position)."
.It "da	bool		Display may be retained above the screen."
.It "dB	num	(o)	Milliseconds of"
.Sy \&bs
delay needed (default 0).
.It "db	bool		Display may be retained below the screen."
.It "DC	str	(NP*)	Delete"
.Em n
characters.
.It "dC	num	(o)	Milliseconds of"
.Sy \&cr
delay needed (default 0).
.It "dc	str	(P*)	Delete character."
.It "dF	num	(o)	Milliseconds of"
.Sy \&ff
delay needed (default 0).
.It "DL	str	(NP*)	Delete"
.Ar n
lines.
.It "dl	str	(P*)	Delete line."
.It "dm	str		Enter delete mode."
.It "dN	num	(o)	Milliseconds of
.Sy \&nl
delay needed (default 0).
.It "DO	str	(NP*)	Move cursor down:
.Ar n
lines.
.It "do	str		Down one line."
.It "ds	str		Disable status line."
.It "dT	num	(o)	Milliseconds of horizontal tab delay needed (default 0)."
.It "dV	num	(o)	Milliseconds of vertical tab delay needed (default 0)."
.It "ec	str	(NP)	Erase"
.Ar n
characters.
.It "ed	str		End delete mode."
.It "ei	str		End insert mode."
.It "eo	bool		Can erase overstrikes with a blank."
.It "EP	bool	(o)	Even parity."
.It "es	bool		Escape can be used on the status line."
.It "ff	str	(P*)	Hardcopy terminal page eject."
.It "fs	str		Return from status line."
.It "gn	bool		Generic line type, for example dialup, switch)."
.It "hc	bool		Hardcopy terminal."
.It "HD	bool	(o)	Half-duplex."
.It "hd	str		Half-line down (forward 1/2 linefeed)."
.It "ho	str	(P)	Home cursor."
.It "hs	bool		Has extra"
.Dq status line .
.It "hu	str		Half-line up (reverse 1/2 linefeed)."
.It "hz	bool		Cannot print ``~'' (Hazeltine)."
.It "i1-i3	str		Terminal initialization strings"
.Pf ( Xr terminfo
only)
.It "IC	str	(NP*)	Insert"
.Ar n
blank characters.
.It "ic	str	(P*)	Insert character."
.It "if	str		Name of file containing initialization string."
.It "im	str		Enter insert mode."
.It "in	bool		Insert mode distinguishes nulls."
.It "iP	str		Pathname of program for initialization"
.Pf ( Xr terminfo
only).
.It "ip	str	(P*)	Insert pad after character inserted."
.It "is	str		Terminal initialization string"
.Pf ( Nm termcap
only).
.It "it	num		Tabs initially every"
.Ar n
positions.
.It "K1	str		Sent by keypad upper left."
.It "K2	str		Sent by keypad center."
.It "K3	str		Sent by keypad upper right."
.It "K4	str		Sent by keypad lower left."
.It "K5	str		Sent by keypad lower right."
.It "k0-k9	str		Sent by function keys 0-9."
.It "kA	str		Sent by insert-line key."
.It "ka	str		Sent by clear-all-tabs key."
.It "kb	str		Sent by backspace key."
.It "kC	str		Sent by clear-screen or erase key."
.It "kD	str		Sent by delete-character key."
.It "kd	str		Sent by down-arrow key."
.It "kE	str		Sent by clear-to-end-of-line key."
.It "ke	str		Out of"
.Dq keypad transmit
mode.
.It "kF	str		Sent by scroll-forward/down key."
.It "kH	str		Sent by home-down key."
.It "kh	str		Sent by home key."
.It "kI	str		Sent by insert-character or enter-insert-mode key."
.It "kL	str		Sent by delete-line key."
.It "kl	str		Sent by left-arrow key."
.It "kM	str		Sent by insert key while in insert mode."
.It "km	bool		Has a"
.Dq meta
key (shift, sets parity bit).
.It "kN	str		Sent by next-page key."
.It "kn	num	(o)	Number of function"
.Pq Sy \&k\&0 Ns \- Sy \&k\&9
keys (default 0).
.It "ko	str	(o)	Termcap entries for other non-function keys."
.It "kP	str		Sent by previous-page key."
.It "kR	str		Sent by scroll-backward/up key."
.It "kr	str		Sent by right-arrow key."
.It "kS	str		Sent by clear-to-end-of-screen key."
.It "ks	str		Put terminal in"
.Dq keypad transmit
mode.
.It "kT	str		Sent by set-tab key."
.It "kt	str		Sent by clear-tab key."
.It "ku	str		Sent by up-arrow key."
.It "l0-l9	str		Labels on function keys if not"
.Dq \&f Ns Em n .
.It "LC	bool	(o)	Lower-case only."
.It "LE	str	(NP)	Move cursor left"
.Ar n
positions.
.It "le	str	(P)	Move cursor left one position."
.It "li	num		Number of lines on screen or page (See"
.Sx BUGS
section below)
.It "ll	str		Last line, first column
.It "lm	num		Lines of memory if >" Sy \&li
(0 means varies).
.It "ma	str	(o)	Arrow key map (used by"
.Xr \&vi
version 2 only).
.It "mb	str		Turn on blinking attribute."
.It "md	str		Turn on bold (extra bright) attribute."
.It "me	str		Turn off all attributes."
.It "mh	str		Turn on half-bright attribute."
.It "mi	bool		Safe to move while in insert mode."
.It "mk	str		Turn on blank attribute (characters invisible)."
.It "ml	str	(o)	Memory lock on above cursor."
.It "mm	str		Turn on"
.Dq meta mode
(8th bit).
.It "mo	str		Turn off"
.Dq meta mode .
.It "mp	str		Turn on protected attribute."
.It "mr	str		Turn on reverse-video attibute."
.It "ms	bool		Safe to move in standout modes."
.It "mu	str	(o)	Memory unlock (turn off memory lock)."
.It "nc	bool	(o)	No correctly-working"
.Sy \&cr
(Datamedia 2500, Hazeltine 2000).
.It "nd	str		Non-destructive space (cursor right)."
.It "NL	bool	(o)" Ta Sy \&\en No "is newline, not line feed."
.It "nl	str	(o)	Newline character if not" Sy \en .
.It "ns	bool	(o)	Terminal is a" Tn CRT No "but doesn't scroll."
.It "nw	str	(P)	Newline (behaves like"
.Sy \&cr
followed by
.Sy \&do ).
.It "OP	bool	(o)	Odd parity."
.It "os	bool		Terminal overstrikes."
.It "pb	num		Lowest baud where delays are required."
.It "pc	str		Pad character (default" Tn NUL ).
.It "pf	str		Turn off the printer."
.It "pk	str		Program function key"
.Em n
to type string
.Em s
.Pf ( Xr terminfo
only).
.It "pl	str		Program function key"
.Em n
to execute string
.Em s
.Pf ( Xr terminfo
only).
.It "pO	str	(N)	Turn on the printer for"
.Em n
bytes.
.It "po	str		Turn on the printer."
.It "ps	str		Print contents of the screen."
.It "pt	bool	(o)	Has hardware tabs (may need to be set with"
.Sy \&is ).
.It "px	str		Program function key"
.Em n
to transmit string
.Em s
.Pf ( Xr terminfo
only).
.It "r1-r3	str		Reset terminal completely to sane modes"
.Pf ( Xr terminfo
only).
.It "rc	str	(P)	Restore cursor to position of last"
.Sy \&sc .
.It "rf	str		Name of file containing reset codes."
.It "RI	str	(NP)	Move cursor right"
.Em n
positions.
.It "rp	str	(NP*)	Repeat character"
.Em c n
times.
.It "rs	str		Reset terminal completely to sane modes"
.Pf ( Nm termcap
only).
.It "sa	str	(NP)	Define the video attributes."
.It "sc	str	(P)	Save cursor position."
.It "se	str		End standout mode."
.It "SF	str	(NP*)	Scroll forward"
.Em n
lines.
.It "sf	str	(P)	Scroll text up."
.It "sg	num		Number of garbage chars left by"
.Sy \&so
or
.Sy \&se
(default 0).
.It "so	str		Begin standout mode."
.It "SR	str	(NP*)	Scroll backward"
.Em n
lines.
.It "sr	str	(P)	Scroll text down."
.It "st	str		Set a tab in all rows, current column."
.It "ta	str	(P)	Tab to next 8-position hardware tab stop."
.It "tc	str		Entry of similar terminal \- must be last."
.It "te	str		String to end programs that use"
.Nm termcap .
.It "ti	str		String to begin programs that use"
.Nm termcap .
.It "ts	str	(N)	Go to status line, column"
.Em n .
.It "UC	bool	(o)	Upper-case only."
.It "uc	str		Underscore one character and move past it."
.It "ue	str		End underscore mode."
.It "ug	num		Number of garbage chars left by"
.Sy \&us
or
.Sy \&ue
(default 0).
.It "ul	bool		Underline character overstrikes."
.It "UP	str	(NP*)	Move cursor up"
.Em n
lines.
.It "up	str		Upline (cursor up)."
.It "us	str		Start underscore mode."
.It "vb	str		Visible bell (must not move cursor)."
.It "ve	str		Make cursor appear normal (undo"
.Sy \&vs Ns / Sy \&vi ) .
.It "vi	str		Make cursor invisible."
.It "vs	str		Make cursor very visible."
.It "vt	num		Virtual terminal number (not supported on all systems)."
.It "wi	str	(N)	Set current window."
.It "ws	num		Number of columns in status line."
.It "xb	bool		Beehive"
.Pf ( "f1=" Dv ESC ,
.Pf "f2=" Sy \&^C ) .
.It "xn	bool		Newline ignored after 80 cols (Concept)."
.It "xo	bool		Terminal uses xoff/xon"
.Pq Dv DC3 Ns / Ns Dv DC1
handshaking.
.It "xr	bool	(o)	Return acts like"
.Sy "ce cr nl"
(Delta Data).
.It "xs	bool		Standout not erased by overwriting (Hewlett-Packard)."
.It "xt	bool		Tabs ruin, magic"
.SY \&so
char (Teleray 1061).
.It "xx	bool	(o)	Tektronix 4025 insert-line."
.El
.Ss A Sample Entry
The following entry, which describes the Concept\-100, is among the more
complex entries in the
.Nm termcap
file as of this writing.
.Pp
.Bd -literal
ca\||\|concept100\||\|c100\||\|concept\||\|c104\||\|concept100-4p\||\|HDS Concept\-100:\e
	:al=3*\eE^R:am:bl=^G:cd=16*\eE^C:ce=16\eE^U:cl=2*^L:cm=\eEa%+ %+ :\e
	:co#80:.cr=9^M:db:dc=16\eE^A:dl=3*\eE^B:do=^J:ei=\eE\e200:eo:im=\eE^P:in:\e
	:ip=16*:is=\eEU\eEf\eE7\eE5\eE8\eEl\eENH\eEK\eE\e200\eEo&\e200\eEo\e47\eE:k1=\eE5:\e
	:k2=\eE6:k3=\eE7:kb=^h:kd=\eE<:ke=\eEx:kh=\eE?:kl=\eE>:kr=\eE=:ks=\eEX:\e
	:ku=\eE;:le=^H:li#24:mb=\eEC:me=\eEN\e200:mh=\eEE:mi:mk=\eEH:mp=\eEI:\e
	:mr=\eED:nd=\eE=:pb#9600:rp=0.2*\eEr%.%+ :se=\eEd\eEe:sf=^J:so=\eEE\eED:\e
	:.ta=8\et:te=\eEv    \e200\e200\e200\e200\e200\e200\eEp\er\en:\e
	:ti=\eEU\eEv  8p\eEp\er:ue=\eEg:ul:up=\eE;:us=\eEG:\e
	:vb=\eEk\e200\e200\e200\e200\e200\e200\e200\e200\e200\e200\e200\e200\e200\e200\eEK:\e
	:ve=\eEw:vs=\eEW:vt#8:xn:\e
	:bs:cr=^M:dC#9:dT#8:nl=^J:ta=^I:pt:
.Ed
.Pp
Entries may continue onto multiple lines by giving a \e as the last
character of a line, and empty fields
may be included for readability (here between the last field on a line
and the first field on the next).
Comments may be included on lines beginning with
.Dq # .
.Ss Types of Capabilities
Capabilities in
.Nm termcap
are of three types: Boolean capabilities,
which indicate particular features that the terminal has;
numeric capabilities,
giving the size of the display or the size of other attributes;
and string capabilities,
which give character sequences that can be used to perform particular
terminal operations.
All capabilities have two-letter codes.
For instance, the fact that
the Concept has
.Em automatic margins
(an automatic return and linefeed
when the end of a line is reached) is indicated by the Boolean capability
.Sy \&am .
Hence the description of the Concept includes
.Sy \&am .
.Pp
Numeric capabilities are followed by the character `#' then the value.
In the example above
.Sy \&co ,
which indicates the number of columns the display has,
gives the value `80' for the Concept.
.Pp
Finally, string-valued capabilities, such as
.Sy \&ce
(clear-to-end-of-line
sequence) are given by the two-letter code, an `=', then a string
ending at the next following `:'.
A delay in milliseconds may appear after
the `=' in such a capability,
which causes padding characters to be supplied by
.Xr tputs
after the remainder of the string is sent to provide this delay.
The delay can be either a number,
such as `20', or a number followed by
an `*',
such as `3*'.
An `*' indicates that the padding required is proportional
to the number of lines affected by the operation, and the amount given is
the per-affected-line padding required.
(In the case of insert-character,
the factor is still the number of
.Em lines
affected;
this is always 1 unless the terminal has
.Sy \&in
and the software uses it.)
When an `*' is specified, it is sometimes useful to give a delay of the form
`3.5' to specify a delay per line to tenths of milliseconds.
(Only one decimal place is allowed.)
.Pp
A number of escape sequences are provided in the string-valued capabilities
for easy encoding of control characters there.
.Sy \&\eE
maps to an
.Dv ESC
character,
.Sy \&^X
maps to a control-X for any appropriate X,
and the sequences
.Sy \&\en
.Sy \&\er
.Sy \&\et
.Sy \&\eb
.Sy \&\ef
map to linefeed, return, tab, backspace, and formfeed, respectively.
Finally, characters may be given as three octal digits after a
.Sy \&\e ,
and the characters
.Sy \&^
and
.Sy \&\e
may be given as
.Sy \&\e^
and
.Sy \&\e\e .
If it is necessary to place a
.Sy \&:
in a capability it must be escaped in
octal as
.Sy \&\e072 .
If it is necessary to place a
.Dv NUL
character in a string capability it
must be encoded as
.Sy \&\e200 .
(The routines that deal with
.Nm termcap
use C strings and strip the high bits of the output very late, so that
a
.Sy \&\e200
comes out as a
.Sy \&\e000
would.)
.Pp
Sometimes individual capabilities must be commented out.
To do this, put a period before the capability name.
For example, see the first
.Sy \&cr
and
.Sy \&ta
in the example above.
.Ss Preparing Descriptions
The most effective way to prepare a terminal description is by imitating
the description of a similar terminal in
.Nm termcap
and to build up a description gradually, using partial descriptions
with
.Xr \&vi
to check that they are correct.
Be aware that a very unusual terminal may expose deficiencies in
the ability of the
.Nm termcap
file to describe it
or bugs in
.Xr \&vi .
To easily test a new terminal description you are working on
you can put it in your home directory in a file called
.Pa .termcap
and programs will look there before looking in
.Pa /usr/share/misc/termcap .
You can also set the environment variable
.Ev TERMPATH
to a list of absolute file pathnames (separated by spaces or colons),
one of which contains the description you are working on,
and programs will search them in the order listed, and nowhere else.
See
.Xr termcap 3 .
The
.Ev TERMCAP
environment variable is usually set to the
.Nm termcap
entry itself
to avoid reading files when starting up a program.
.Pp
To get the padding for insert-line right
(if the terminal manufacturer did not document it),
a severe test is to use
.Xr \&vi
to edit
.Pa /etc/passwd
at 9600 baud, delete roughly 16 lines from the middle of the screen,
then hit the `u' key several times quickly.
If the display messes up, more padding is usually needed.
A similar test can be used for insert-character.
.Ss Basic Capabilities
The number of columns on each line of the display is given by the
.Sy \&co
numeric capability.
If the display is a
.Tn CRT ,
then the
number of lines on the screen is given by the
.Sy \&li
capability.
If the display wraps around to the beginning of the next line when
the cursor reaches the right margin, then it should have the
.Sy \&am
capability.
If the terminal can clear its screen,
the code to do this is given by the
.Sy \&cl
string capability.
If the terminal overstrikes
(rather than clearing the position when a character is overwritten),
it should have the
.Sy \&os
capability.
If the terminal is a printing terminal,
with no soft copy unit,
give it both
.Sy \&hc
and
.Sy \&os .
.Pf ( Sy \&os
applies to storage scope terminals,
such as the Tektronix 4010 series,
as well as to hard copy and
.Tn APL
terminals.)
If there is a code to move the cursor to the left edge of the current row,
give this as
.Sy \&cr .
(Normally this will be carriage-return,
.Sy \&^M . )
If there is a code to produce an audible signal (bell, beep,
etc. ) ,
give this as
.Sy \&bl .
.Pp
If there is a code (such as backspace)
to move the cursor one position to the left,
that capability should be given as
.Sy \&le .
Similarly,
codes to move to the right, up, and down
should be given as
.Sy \&nd ,
.Sy \&up ,
and
.Sy \&do ,
respectively.
These
.Em local cursor motions
should not alter the text they pass over;
for example, you would not normally use
.Dq nd=\ \&
unless the terminal has the
.Sy \&os
capability,
because the space would erase the character moved over.
.Pp
A very important point here is that the local cursor motions encoded
in
.Nm termcap
have undefined behavior at the left and top edges of a
.Tn CRT
display.
Programs should never attempt to backspace around the left edge,
unless
.Sy \&bw
is given, and never attempt to go up off the top
using local cursor motions.
.Pp
In order to scroll text up,
a program goes to the bottom left corner of the screen and sends the
.Sy \&sf
(index) string.
To scroll text down,
a program goes to the top left corner of the screen and sends the
.Sy \&sr
(reverse index) string.
The strings
.Sy \&sf
and
.Sy \&sr
have undefined behavior
when not on their respective corners of the screen.
Parameterized versions of the scrolling sequences are
.Sy \&SF
and
.Sy \&SR ,
which have the same semantics as
.Sy \&sf
and
.Sy \&sr
except that they take one parameter
and scroll that many lines.
They also have undefined behavior
except at the appropriate corner of the screen.
.Pp
The
.Sy \&am
capability tells whether the cursor sticks at the right
edge of the screen when text is output there,
but this does not necessarily apply to
.Sy \&nd
from the last column.
Leftward local motion is defined from the left edge only when
.Sy \&bw
is given; then an
.Sy \&le
from the left edge will move to the right edge of the previous row.
This is useful for drawing a box around the edge of the screen,
for example.
If the terminal has switch-selectable automatic margins,
the
.Nm termcap
description usually assumes that this feature is on,
.Em i.e . ,
.Sy \&am .
If the terminal has a command
that moves to the first column of the next line,
that command can be given as
.Sy \&nw
(newline).
It is permissible for this to clear the remainder of the current line,
so if the terminal has no correctly-working
.Tn \&CR
and
.Tn \&LF
it may still be possible to craft a working
.Sy \&nw
out of one or both of them.
.Pp
These capabilities suffice to describe hardcopy and
.Dq glass-tty
terminals.
Thus the Teletype model 33 is described as
.Bd -literal -offset indent
T3\||\|tty33\||\|33\||\|tty\||\|Teletype model 33:\e
	:bl=^G:co#72:cr=^M:do=^J:hc:os:
.Ed
.Pp
and the Lear Siegler
.Tn ADM Ns \-3
is described as
.Bd -literal -offset indent
l3\||\|adm3\||\|3\||\|LSI \s-1ADM\s0-3:\e
:am:bl=^G:cl=^Z:co#80:cr=^M:do=^J:le=^H:li#24:sf=^J:
.Ed
.Ss Parameterized Strings
Cursor addressing and other strings requiring parameters
are described by a
parameterized string capability, with
.Xr printf 3 Ns \-like
escapes
.Sy \&%x
in it,
while other characters are passed through unchanged.
For example, to address the cursor the
.Sy \&cm
capability is given, using two parameters: the row and column to move to.
(Rows and columns are numbered from zero and refer to the physical screen
visible to the user, not to any unseen memory.
If the terminal has memory-relative cursor addressing,
that can be indicated by an analogous
.Sy \&CM
capability.)
.Pp
The
.Sy \&%
encodings have the following meanings:
.Bl -column xxxxx
.It "%%	output `%'"
.It "%d	output value as in"
.Xr printf
%d
.It "%2	output value as in"
.Xr printf
%2d
.It "%3	output value as in"
.Xr printf
%3d
.It "%.	output value as in"
.Xr printf
%c
.It "%+" Ns Em x Ta No add
.Em x
to value, then do %.
.It "%>" Ns Em \&xy Ta No if
value >
.Em x
then add
.Em y ,
no output
.It "%r	reverse order of two parameters, no output"
.It "%i	increment by one, no output"
.It "%n	exclusive-or all parameters with 0140 (Datamedia 2500)"
.It "%B" Ta Tn BCD No "(16*(value/10)) + (value%10), no output"
.It "%D	Reverse coding (value \- 2*(value%16)), no output (Delta Data)."
.El
.Pp
Consider the Hewlett-Packard 2645, which, to get to row 3 and column 12, needs
to be sent
.Dq \eE&a12c03Y
padded for 6 milliseconds.
Note that the order
of the row and column coordinates is reversed here
and that the row and column
are sent as two-digit integers.
Thus its
.Sy \&cm
capability is
.Dq Li cm=6\eE&%r%2c%2Y .
.Pp
The Datamedia 2500 needs the current row and column sent
encoded in binary using
.Dq \&%. .
Terminals that use
.Dq \&%.
need to be able to
backspace the cursor
.Po Sy \&le Pc
and to move the cursor up one line on the screen
.Po Sy \&up Pc .
This is necessary because it is not always safe to transmit
.Sy \&\en ,
.Sy \&^D ,
and
.Sy \&\er ,
as the system may change or discard them.
(Programs using
.Nm termcap
must set terminal modes so that tabs are not expanded, so
.Sy \&\et
is safe to send.
This turns out to be essential for the Ann Arbor 4080.)
.Pp
A final example is the Lear Siegler
.Tn ADM Ns \-3a,
which offsets row and column
by a blank character, thus
.Dq Li cm=\eE=%+ %+\ \& .
.Pp
Row or column absolute cursor addressing
can be given as single parameter capabilities
.Sy \&ch
(horizontal position absolute) and
.Sy \&cv
(vertical position absolute).
Sometimes these are shorter than the more general two-parameter sequence
(as with the Hewlett-Packard 2645) and can be used in preference to
.Sy \&cm .
If there are parameterized local motions
.Pf ( Em e.g . ,
move
.Ar n
positions to the right)
these can be given as
.Sy \&DO ,
.Sy \&LE ,
.Sy \&RI ,
and
.Sy \&UP
with a single parameter indicating how many positions to move.
These are primarily useful if the terminal does not have
.Sy \&cm ,
such as the Tektronix 4025.
.Ss Cursor Motions
.Pp
If the terminal has a fast way to home the cursor
(to the very upper left corner of the screen), this can be given as
.Sy \&ho .
Similarly, a fast way of getting to the lower left-hand corner
can be given as
.Sy \&ll ;
this may involve going up with
.Sy \&up
from the home position,
but a program should never do this itself (unless
.Sy \&ll
does), because it can
make no assumption about the effect of moving up from the home position.
Note that the home position is the same as
cursor address (0,0): to the top left corner of the screen, not of memory.
(Therefore, the
.Dq \eEH
sequence on Hewlett-Packard terminals
cannot be used for
.Sy \&ho . )
.Ss Area Clears
If the terminal can clear from the current position to the end of the
line, leaving the cursor where it is, this should be given as
.Sy \&ce .
If the terminal can clear from the current position to the end of the
display, this should be given as
.Sy \&cd .
.Sy \&cd
must only be invoked from the first column of a line.
(Therefore,
it can be simulated by a request to delete a large number of lines,
if a true
.Sy \&cd
is not available.)
.Ss Insert/Delete Line
If the terminal can open a new blank line
before the line containing the cursor,
this should be given as
.Sy \&al ;
this must be invoked only from the first
position of a line.
The cursor must then appear at the left of the newly blank line.
If the terminal can delete the line that the cursor is on, this
should be given as
.Sy \&dl ;
this must only be used from the first position on
the line to be deleted.
Versions of
.Sy \&al
and
.Sy \&dl
which take a single parameter
and insert or delete that many lines
can be given as
.Sy \&AL
and
.Sy \&DL .
If the terminal has a settable scrolling region
(like the VT100),
the command to set this can be described with the
.Sy \&cs
capability,
which takes two parameters: the top and bottom lines of the scrolling region.
The cursor position is, alas, undefined after using this command.
It is possible to get the effect of insert or delete line
using this command \(em the
.Sy \&sc
and
.Sy \&rc
(save and restore cursor) commands are also useful.
Inserting lines at the top or bottom of the screen can also be done using
.Sy \&sr
or
.Sy \&sf
on many terminals without a true insert/delete line,
and is often faster even on terminals with those features.
.Pp
If the terminal has the ability to define a window as part of memory
which all commands affect, it should be given as the parameterized string
.Sy \&wi .

gnu/usr.bin/perl/lib/Pod/Text/Termcap.pm
# Pod::Text::Termcap -- Convert POD data to ASCII text with format escapes.
# $Id: Termcap.pm,v 1.1 2000/04/06 16:09:48 millert Exp $
#
# Copyright 1999 by Russ Allbery <rra@stanford.edu>
#
# This program is free software; you can redistribute it and/or modify it
# under the same terms as Perl itself.
#
# This is a simple subclass of Pod::Text that overrides a few key methods to
# output the right termcap escape sequences for formatted text on the
# current terminal type.

############################################################################
# Modules and declarations
############################################################################

package Pod::Text::Termcap;

require 5.004;

use Pod::Text ();
use POSIX ();
use Term::Cap;

use strict;
use vars qw(@ISA $VERSION);

@ISA = qw(Pod::Text);

# Use the CVS revision of this file as its version number.
($VERSION = (split (' ', q$Revision: 1.1 $ ))[1]) =~ s/\.(\d)$/.0$1/;


############################################################################
# Overrides
############################################################################

# In the initialization method, grab our terminal characteristics as well as
# do all the stuff we normally do.
sub initialize {
    my $self = shift;

    # The default Term::Cap path won't work on Solaris.
    $ENV{TERMPATH} = "$ENV{HOME}/.termcap:/etc/termcap"
        . ":/usr/share/misc/termcap:/usr/share/lib/termcap";

    my $termios = POSIX::Termios->new;
    $termios->getattr;
    my $ospeed = $termios->getospeed;
    my $term = Tgetent Term::Cap { TERM => undef, OSPEED => $ospeed };
    $$self{BOLD} = $$term{_md} or die 'BOLD';
    $$self{UNDL} = $$term{_us} or die 'UNDL';
    $$self{NORM} = $$term{_me} or die 'NORM';

    unless (defined $$self{width}) {
        $$self{width} = $ENV{COLUMNS} || $$term{_co} || 78;
        $$self{width} -= 2;
    }

    $self->SUPER::initialize;
}

# Make level one headings bold.
sub cmd_head1 {
    my $self = shift;
    local $_ = shift;
    s/\s+$//;
    $self->SUPER::cmd_head1 ("$$self{BOLD}$_$$self{NORM}");
}

# Make level two headings bold.
sub cmd_head2 {
    my $self = shift;
    local $_ = shift;
    s/\s+$//;
    $self->SUPER::cmd_head2 ("$$self{BOLD}$_$$self{NORM}");
}

# Fix up B<> and I<>.  Note that we intentionally don't do F<>.
sub seq_b { my $self = shift; return "$$self{BOLD}$_[0]$$self{NORM}" }
sub seq_i { my $self = shift; return "$$self{UNDL}$_[0]$$self{NORM}" }

# Override the wrapping code to igore the special sequences.
sub wrap {
    my $self = shift;
    local $_ = shift;
    my $output = '';
    my $spaces = ' ' x $$self{MARGIN};
    my $width = $$self{width} - $$self{MARGIN};
    my $code = "(?:\Q$$self{BOLD}\E|\Q$$self{UNDL}\E|\Q$$self{NORM}\E)";
    while (length > $width) {
        if (s/^((?:$code?[^\n]){0,$width})\s+//
            || s/^((?:$code?[^\n]){$width})//) {
            $output .= $spaces . $1 . "\n";
        } else {
            last;
        }
    }
    $output .= $spaces . $_;
    $output =~ s/\s+$/\n\n/;
    $output;
}


############################################################################
# Module return value and documentation
############################################################################

1;
__END__

=head1 NAME

Pod::Text::Color - Convert POD data to ASCII text with format escapes

=head1 SYNOPSIS

    use Pod::Text::Termcap;
    my $parser = Pod::Text::Termcap->new (sentence => 0, width => 78);

    # Read POD from STDIN and write to STDOUT.
    $parser->parse_from_filehandle;

    # Read POD from file.pod and write to file.txt.
    $parser->parse_from_file ('file.pod', 'file.txt');

=head1 DESCRIPTION

Pod::Text::Termcap is a simple subclass of Pod::Text that highlights output
text using the correct termcap escape sequences for the current terminal.
Apart from the format codes, it in all ways functions like Pod::Text.  See
L<Pod::Text> for details and available options.

=head1 SEE ALSO

L<Pod::Text|Pod::Text>, L<Pod::Parser|Pod::Parser>

=head1 AUTHOR

Russ Allbery E<lt>rra@stanford.eduE<gt>.

=cut
